keps/sig-node/5304-dra-attributes-downward-api/README.md

```diff
@@ -0,0 +1,830 @@
+
+Risk: API surface expansion could make the Downward API overly complex.
+Mitigation: Keep the API minimal and type-safe with a dedicated selector; no arbitrary JSONPath. Require feature gate enablement in alpha to gather feedback before expanding.
+
+Risk: Compatibility with future DRA changes.
+Mitigation: Implementation is decoupled from DRA driver interfaces. Changes to DRA object structure are handled in the kubelet controller. Attribute names are standardized and versioned.
+
+## Design Details
+
+### API Changes
+
+#### New Downward API Selector: ResourceSliceAttributeSelector
+
+A new typed selector is introduced in `core/v1` that can be used in both environment variable sources and projected downward API volume items:
+
+```go
+// ResourceSliceAttributeSelector selects a DRA-resolved device attribute for a given claim+request.
+// +featureGate=DRADownwardDeviceAttributes
+// +structType=atomic
+type ResourceSliceAttributeSelector struct {
+    // ClaimName must match pod.spec.resourceClaims[].name.
+    // +required
+    ClaimName string `json:"claimName"`
+    
+    // RequestName must match the corresponding ResourceClaim.spec.devices.requests[].name.
+    // +required
+    RequestName string `json:"requestName"`
+    
+    // Attribute specifies which device attribute to expose from the ResourceSlice.
+    // The attribute name must be present in the ResourceSlice's device attributes.
+    // +required
+    Attribute string `json:"attribute"`
+}
+
+// In core/v1 EnvVarSource:
+type EnvVarSource struct {
+    // ...existing fields...
+    
+    // ResourceSliceAttributeRef selects a DRA device attribute for a given claim+request.
+    // +featureGate=DRADownwardDeviceAttributes
+    // +optional
+    ResourceSliceAttributeRef *ResourceSliceAttributeSelector `json:"resourceSliceAttributeRef,omitempty"`
+}
+
+// (Projected volume support deferred to beta)
+```
+
+Validation:
+- Enforce exactly one of `fieldRef`, `resourceFieldRef`, or `resourceSliceAttributeRef` in both env and volume items
+- Validate `claimName` and `requestName` against DNS label rules
+- No API-level enumeration of attribute names; kubelet resolves attributes that exist in the matching `ResourceSlice` at runtime
+
+
+####
+
+### Kubelet Implementation
+
+The kubelet runs a local DRA attributes controller that:
+
+1. Watches Pods: Identifies Pods on the node with `pod.spec.resourceClaims` and tracks their `pod.status.resourceClaimStatuses` to discover generated ResourceClaim names
+2. Watches ResourceClaims: For each relevant claim, reads `status.allocation.devices.results[*]` and maps entries by request name
+3. Watches ResourceSlices: Resolves standardized attributes from `spec.devices[*].attributes` for the matching device name
+4. Maintains Cache: Keeps a per-Pod map of `(claimName, requestName) -> {attribute: value}` with a readiness flag
```

Comment:
Can you please elaborate on the implementation?
Does it open a new watch per ResourceClaim and ResourceSlice?
The Kubelet already does a Get on the NodePrepareResources hook and passes the claim to the driver